### ISSUE TITLE
"cargo miri test" ignores "required-features"

### ISSUE THREAD

**Issue by alecmocatta** [rust-lang/miri#1512]
`cargo test` compiles and runs only tests that aren't disabled via `required-features = ["foo"]` (or `test = false`) in Cargo.toml.

`cargo miri test` tries to compile and run all tests, resulting in either `` target `test_foo` in package `xyz` requires the features: `foo` `` or arbitrary errors from explicitly disabled tests.

A fix for this would probably check for the args `--no-default-features` `--features="..."` and `--all-features`, pass them to [`cargo_metadata::MetadataCommand`](https://docs.rs/cargo_metadata/0.11.1/cargo_metadata/struct.MetadataCommand.html#method.features), read [`Node::features`](https://docs.rs/cargo_metadata/0.11.1/cargo_metadata/struct.Node.html#structfield.features), and then filter the targets returned by [`list_targets()`](https://github.com/rust-lang/miri/blob/ca9e988bdb24838e5e23e40bcc936fd35ce7dede/cargo-miri/bin.rs#L119-L161).

**Comment by RalfJung** [comment]
Thanks for the report!

The part about `test = false` is a duplicate of https://github.com/rust-lang/miri/issues/700, which is blocked on https://github.com/oli-obk/cargo_metadata/issues/114. So let's use this issue for the `required-features` problem, which I was not aware of.

Really the best thing would be if we didn't have to re-implement the entire binary selection of cargo, which is tracked separately in https://github.com/rust-lang/miri/issues/739.
